cmd/root.go

```diff
@@ -529,30 +511,169 @@ status code.`)
+	v, err := getViper(c)
+	if err != nil {
+		return err
+	}
+
+	_ = v.BindPFlags(c.Flags())
+
+	c.Flags().VisitAll(func(f *pflag.Flag) {
+		// Override any unset flags with Viper values to use the pflags
+		// object as a single source of truth.
 		if !f.Changed && v.IsSet(f.Name) {
 			val := v.Get(f.Name)
-			_ = localFlags.Set(f.Name, fmt.Sprintf("%v", val))
+			_ = c.Flags().Set(f.Name, fmt.Sprintf("%v", val))
 		}
 	})
-	globalFlags.VisitAll(func(f *pflag.Flag) {
-		if !f.Changed && v.IsSet(f.Name) {
-			val := v.Get(f.Name)
-			_ = globalFlags.Set(f.Name, fmt.Sprintf("%v", val))
+
+	return nil
+}
+
+func loadArgs(c *Command, args []string) error {
+	v, err := getViper(c)
+	if err != nil {
+		return err
+	}
+
+	// If args is not already populated, try to read from the environment.
+	if len(args) == 0 {
+		args = instanceFromEnv(args)
+	}
+
+	// If no environment args are present, try to read from the config file.
+	if len(args) == 0 {
+		args = instanceFromConfigFile(v)
+	}
+
+	// Handle logger separately from config
+	if c.conf.StructuredLogs {
+		c.logger, c.cleanup = log.NewStructuredLogger(c.conf.Quiet)
+	}
+
+	if c.conf.Quiet {
+		c.logger = log.NewStdLogger(io.Discard, os.Stderr)
+	}
+
+	err = parseConfig(c, c.conf, args)
+	if err != nil {
+		return err
+	}
+
+	// The arguments are parsed. Usage is no longer needed.
+	c.SilenceUsage = true
+
+	// Errors will be handled by logging from here on.
+	c.SilenceErrors = true
+
+	return nil
+}
+
+func getViper(c *Command) (*viper.Viper, error) {
+	v := viper.New()
+
+	if c.conf.ConfigFilepath != "" {
+		// Setup Viper configuration file. Viper will attempt to load configuration
+		// from the specified file if it exists. Otherwise, Viper will source all
+		// configuration from flags and then environment variables.
+		ext := filepath.Ext(c.conf.ConfigFilepath)
+
+		if ext == "" {
+			return nil, newBadCommandError(fmt.Sprintf("config file %v has no extension, "+
+				"requires TOML, YAML or JSON format", c.conf.ConfigFilepath))
 		}
-	})
 
-	return c
+		if ext != ".toml" && ext != ".yaml" && ext != ".yml" && ext != ".json" {
+			return nil, newBadCommandError(fmt.Sprintf("config file %v has an invalid extension, "+
+				"requires TOML, YAML or JSON format", c.conf.ConfigFilepath))
+		}
+
+		conf := filepath.Base(c.conf.ConfigFilepath)
+		noExt := strings.ReplaceAll(conf, ext, "")
+		v.SetConfigName(noExt)   // argument must be the name of config file without extension
+		v.SetConfigType(ext[1:]) // setting the config type makes the extension optional
+		v.AddConfigPath(filepath.Dir(c.conf.ConfigFilepath))
+
+		// Attempt to load configuration from a file. If no file is found, assume
+		// configuration is provided by flags or environment variables.
+		if err := v.ReadInConfig(); err != nil {
+			// If the error is a ConfigFileNotFoundError, then ignore it.
+			// Otherwise, report the error to the user.
+			var configFileNotFoundError viper.ConfigFileNotFoundError
```

Comment:
nit: the type tells us what it is, so `cErr`, or something similarly short is better.

Reply:
yes, renamed to cErr.